wdl/GetTruthOverlap.wdl

```diff
@@ -0,0 +1,232 @@
+    String num_samples_file = "num_samples.txt"
+    # if vcf_index is not supplied, try this path automatically:
+    String automatic_vcf_index = vcf + ".tbi"
+
+    runtime {
+        docker: sv_base_docker
+        cpu: 1
+        preemptible: 3
+        max_retries: 0
+        memory: "2 GiB"
+        disks: "local-disk " + disk_gb + " HDD"
+    }
+
+    command <<<
+        set -eu -o pipefail
+
+        INDICES_LIST=~{write_lines(select_all([vcf_index]))}
+        if [ -s $INDICES_LIST ]; then
+            # if index exists at this path, copy it to current working dir
+            if gsutil -q stat ~{automatic_vcf_index}; then
+                gsutil cp ~{automatic_vcf_index} .
+            fi
+        else
+            # vcf index was supplied, symlink it to current working dir
+            read INDEX_FILE < $INDICES_LIST
+            ln -s "$INDEX_FILE" .
+        fi
+
+        export GCS_OAUTH_TOKEN=`gcloud auth application-default print-access-token`
+        bcftools query -l ~{vcf} | wc -w > ~{num_samples_file}
+        # get num records from index. If index is wrong kind or missing, build index on the fly.
+        {
+            bcftools index --nrecords ~{vcf} || {
+                bcftools index -f -t -o "$(basename ~{vcf}).tbi" ~{vcf}
+                bcftools index --nrecords ~{vcf}
+            }
+        } > ~{num_records_file}
+    >>>
+
+    output {
+        Int num_records = read_int(num_records_file)
+        Int num_samples = read_int(num_samples_file)
+        Int num_entries = num_records * num_samples
+    }
+}
+
+
+task GetTruthOverlapTask {
+    input {
+        Array[File] test_vcfs
+        Array[File] truth_vcfs
+        Map[String, File]? vapor_files
+        Float vapor_size_gib
+        Array[File] ped_files
+        File? optimal_overlap_cutoffs
+        Array[String] get_truth_overlap_args = []
+        String sv_utils_docker
+        Int? max_test_records
+        Int? max_truth_records
+    }
+
+    # High disk size for large throughput. A large proportion of run time is loading data from huge VCFs. Disk is cheap.
+    Int disk_gb = round(1000 + size(test_vcfs, "GiB") + size(truth_vcfs, "GiB") + size(ped_files, "GiB")
+                        + size(optimal_overlap_cutoffs, "GiB") + vapor_size_gib)
+    String optimal_overlap_cutoffs_filename = if defined(optimal_overlap_cutoffs)
+        then basename(select_first([optimal_overlap_cutoffs]))
+        else "optimal_overlap_cutoffs.pickle"
+
+    String truth_overlap_info_filename = "truth_overlap.json"
+    String args_str = if length(get_truth_overlap_args) > 0 then sep(" ", get_truth_overlap_args) else ""
+
+    Float mem_baseline = 1.0
+    Float mem_scale = "3.5e-6"
+    Float mem_gb = if (defined(max_test_records) && defined(max_truth_records))
+        then mem_baseline + mem_scale * (select_first([max_test_records]) + select_first([max_truth_records]))
+        else 3.0
+
+    runtime {
+        docker: sv_utils_docker
+        cpu: 12
```

Comment:
The way the task works right now, every sample in the test set is checked for overlaps against the truth set serially (because we only look for overlaps where the test set is called non-REF). The parallelization comes in because the overlap task breaks the genome into chunks (by contig, and where there are gaps within a contig), and the chunks are processed in parallel. Twelve seemed like a reasonable amount of parallelization to ensure that each worker had a reasonable amount of work to do (without any workers finishing early), while still speeding up the wall clock time. I haven't extensively tested, but I doubt it could go higher than the number of chromosomes and still be efficient.

I could in principle get more parallelization by parallelizing over samples and then concatenating those results. That would require some changes to the algorithm though.

Reply:
Fine if it is parallelizing over contigs, as long as it doesn't depend on input size